File: src/CPort/emulator_port.c
#include <stdio.h>
#include <stdlib.h>

#include "emulator_port.h"

int cpu_req(int code, mother_board_t *mother_board, cpu_t *self)
{
	int i = 0;
	int (*instruction)(struct mother_board_s *mother_board, void *dev) = &NUL;

	for (i; (OP_CPU[i]->instruction != NULL || OP_CPU[i]->id != -1) && OP_CPU[i]; ++i) {
		if (OP_CPU[i]->id == code)
			instruction = OP_CPU[i]->instruction;
	}
	return (instruction(mother_board, self));
}

void clock(mother_board_t *mother_board, cpu_t *cpu)
{
	if (cpu->registries->prg_device == mother_board->ram->id)
		cpu->req(mother_board->ram->read(cpu->registries->ptr_prg, mother_board->ram), mother_board, cpu);
	cpu->registries->ptr_prg += 1;
}

int read_ram(int addr, ram_t *ram)
{
	if (!ram)
		return (0x00);
	if (addr < 0x00 || addr >= ram->size)
		return (0x00);
	return ram->memory[addr];
}

int write_ram(int addr, int value, ram_t *ram)
{
	if (!ram)
		return (0x00);
	if (addr < 0x00 || addr >= ram->size)
		return (0x00);
	ram->memory[addr] = value;
	return (0x01);
}

registries_t *create_registries(void)
{
	registries_t *reg = malloc(sizeof(registries_t));

	if (!reg)
		return (NULL);
	reg->x = 0x00;
	reg->y = 0x00;
	reg->temp = 0x00;
	reg->fetched = 0x00;

	reg->ptr_prg = 0x00;
	reg->ptr_mem = 0x00;
	reg->prg_device = 0x01;
	reg->mem_device = 0x01;
	return (reg);
}

cpu_t *create_cpu(void)
{
	cpu_t *cpu = malloc(sizeof(cpu_t));

	if (!cpu)
		return (NULL);
	cpu->req = &cpu_req;
	cpu->id = 0x02;
	cpu->clock = &clock;
	cpu->registries = create_registries();
	return (cpu);
}

ram_t *create_ram(int size)
{
	ram_t *ram = malloc(sizeof(ram_t));

	if (!ram)
		return (NULL);
	ram->memory = malloc(sizeof(int) * size);
	ram->size = size;
	ram->id = 0x01;
	ram->read = &read_ram;
	ram->write = &write_ram;
	if (!ram->memory)
		return (ram);
	for (int i = 0; i < size; ++i)
		ram->memory[i] = 0x00;
	return (ram);
}

mother_board_t *create_mother_board(void)
{
	mother_board_t *mb = malloc(sizeof(mother_board_t));

	if (!mb)
		return (NULL);
	mb->cpu = create_cpu();
	mb->ram = create_ram(8 * 1024);
	return (mb);
}

void destroy_ram(ram_t *ram)
{
	if (!ram)
		return;
	if (ram->memory)
    	free(ram->memory);
	free(ram);
}

void destroy_cpu(cpu_t *cpu)
{
	if (!cpu)
		return;
	if (cpu->registries)
		free(cpu->registries);
	free(cpu);
}

void destroy_mother_board(mother_board_t *mb)
{
	if (!mb)
		return;
	destroy_ram(mb->ram);
	destroy_cpu(mb->cpu);
	free(mb);
}

File: src/JSPort/emulator_port.js
const { get_device } = require("./common.js");

OP = require("./instructions.js");

module.exports = {
    CPU: class {
        constructor() {
            this.id = 0x02;
            this.registries = {
                "x": 0x00,
                "y": 0x00,
                "temp": 0x00,
                "fetched": 0x00,

                "ptr_prg": 0x00,
                "ptr_mem": 0x00,
                "mem_device": 0x01,
                "prg_device": 0x01
            };
            this.instructions = {
                0x00: OP.NULL,
                0x01: OP.ICX,
                0x02: OP.ICY,  
                0x03: OP.ADX,
                0x04: OP.ADY,
                0x05: OP.SUX,
                0x06: OP.IDN,
                0x07: OP.SUY,
                0x08: OP.MUX,
                0x09: OP.MUY,
                0x0A: OP.DIX,
                0x0B: OP.DIY,
                0x0C: OP.MOX,
                0x0D: OP.MOY,
                0x0E: OP.TVF,
                0x0F: OP.END,
                0x10: OP.TVX,
                0x11: OP.TVY,
                0x12: OP.PPX,
                0x13: OP.PPY,
                0x14: OP.MPX,
                0x15: OP.MPY,
                0x16: OP.DMX,
                0x17: OP.DMY,
                0x18: OP.DPX,
                0x19: OP.DPY,
                0x1A: OP.XVY,
                0x1B: OP.YVX,
                0x1C: OP.XVT,
                0x1D: OP.YVT,
                0x1E: OP.XVF,
                0x1F: OP.YVF,
                0x20: OP.MVX,
                0x21: OP.MVY,
                0x22: OP.RDM,
                0x23: OP.TMX,
                0x24: OP.TMY,
                0x25: OP.ANX,
                0x26: OP.ANY,
                0x27: OP.ORX,
                0x28: OP.ORY,
                0x29: OP.XOX,
                0x2A: OP.XOY,
                0x2B: OP.LSX,
                0x2C: OP.LSY,
                0x2D: OP.RSX,
                0x2E: OP.RSY,
                0x2F: OP.EQX,
                0x30: OP.EQY,
                0x31: OP.NQX,
                0x32: OP.NQY,
                0x33: OP.SPX,
                0x34: OP.SPY,
                0x35: OP.INX,
                0x36: OP.INY,
                0x37: OP.IEX,
                0x38: OP.IEY,
                0x39: OP.SEX,
                0x3A: OP.SEY,
                0x3B: OP.NOX,
                0x3C: OP.NOY,
                0x3D: OP.CNE,
                0x3E: OP.HALT,
                0x3F: OP.XPP,
                0x40: OP.YPP,
                0x41: OP.SWX,
                0x42: OP.SWY,
                0x43: OP.REX,
                0x44: OP.REY,
                0x45: OP.XMP,
                0x46: OP.YMP,
                0x47: OP.STX,
                0x48: OP.STY,
                0xFF: OP.LOG          
            };
        }
        send_message(motherboard, instruction) {
            if (!(instruction in this.instructions))
                instruction = 0x00;
            return (this.instructions[instruction](motherboard, this));
        }
        clock(motherboard) {
            let dev = get_device(motherboard, 0x01);
            let instruction = 0x00;

            if (dev != undefined)
                instruction = dev.read(this.registries["ptr_prg"]);
            this.send_message(motherboard, instruction);
            this.registries["ptr_prg"] += 1;
        }
    },
    RAM: class {
        constructor(size) {
            this.id = 0x01;
            this.memory = new Array(size);
            for (let i = 0; i < size; ++i) {
                this.memory[i] = 0x00;
            }
        }
        send_message(motherboard, instruction) {
            return (0x00);
        }
        read(addr) {
            if (addr < 0x00 || addr > this.memory.length)
                return (0x00);
            return (this.memory[addr]);
        }
        write(addr, value) {
            if (addr < 0x00 || addr > this.memory.length)
                return (0x01);
            this.memory[addr] = value;
            return (0x00);
        }
    },
    MotherBoard: class {
        constructor() {
            this.id = 0x00;
            this.connected_dev = [
                new module.exports.RAM(1024),
                new module.exports.CPU()
            ];
        }
    }
}
